### PR TITLE
Implement decomposition for `BitonicSort`

### PR DESCRIPTION
**Pull request by anurudhp**
Decomposition for sorting lists whose size is a power of 2. Partially addresses https://github.com/quantumlib/Qualtran/issues/219

Described briefly in https://www.nature.com/articles/s41534-018-0071-5, circuit based on [wiki](https://en.wikipedia.org/wiki/Bitonic_sorter#:~:text=at%20the%20bottom.-,Alternative%20representation,-%5Bedit%5D).

### CODE REVIEW

**Comment by anurudhp** [comment]
@tanujkhattar @fdmalone ptal

File: qualtran/bloqs/arithmetic/comparison_test.py [hunk 1]
```diff
@@ -230,7 +230,9 @@ def test_greater_than_manual():
     qlt_testing.assert_wire_symbols_match_expected(
         GreaterThanConstant(bitsize, 17), ['In(x)', '⨁(x > 17)']
     )
-    assert cbloq.t_complexity() == t_complexity(LessThanEqual(bitsize, bitsize))
+    assert cbloq.t_complexity() == (
+        t_complexity(LessThanEqual(bitsize, bitsize)) + TComplexity(clifford=1)
+    )
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/comparison_test.py:236]: Can we add a correctness test for the circuit here / tensor contraction / classical simulation?

Comment (anurudhp) [reply qualtran/bloqs/arithmetic/comparison_test.py:236]: added some classical simulation tests. indeed caught a lot of bugs 😅. now works on random lists of size 2-16

Comment (fdmalone) [reply qualtran/bloqs/arithmetic/comparison_test.py:236]: Great, (make sure the tests don't take too long or else @mpharrigan will be on your case :))

File: qualtran/bloqs/arithmetic/sorting.ipynb [hunk 2]
```diff
@@ -36,19 +36,20 @@
    },
    "source": [
     "## `Comparator`\n",
-    "Compare and potentially swaps two n-bit numbers.\n",
+    "Compare and potentially swaps two numbers in {0, ... L - 1}.\n",
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting.ipynb:39]: Why the change from bitsize (which is common in qualtran) to integer range. Is it just more convenient?

File: qualtran/bloqs/arithmetic/sorting.py [hunk 3]
```diff
@@ -11,118 +11,305 @@
...
+    """
+    L: SymbolicInt
+    k: SymbolicInt
+
+    def __attrs_post_init__(self):
+        k = self.k
+        if not is_symbolic(k):
+            assert k >= 1, "length of input lists must be positive"
+            # TODO support non-power-of-two input lengths
+            assert (k & (k - 1)) == 0, "length of input lists must be a power of 2"
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting.py:177]: cute

File: qualtran/bloqs/arithmetic/sorting.ipynb (hunk 2 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.ipynb:39]: papers seem to use "integers in range [n]" etc. often, so I thought this would be more useful. And having `BoundedQUInt` makes it much nicer to support this than earlier.

File: qualtran/bloqs/arithmetic/comparison_test.py (hunk 1 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/comparison_test.py:236]: hahaha yep, total is ~0.5s, should be good

File: qualtran/bloqs/arithmetic/sorting_test.py [hunk 4]
```diff
@@ -27,19 +30,54 @@ def test_bitonic_sort(bloq_autotester):
 
 
 def test_comparator_manual():
-    bloq = Comparator(4)
-    assert bloq.t_complexity().t == 88 - 4
-    with pytest.raises(NotImplementedError):
-        bloq.decompose_bloq()
+    bloq = Comparator(2**4)
+    assert bloq.t_complexity().t == 88 - 4 - 7 * 4
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting_test.py:34]: why the -7 * 4?

File: qualtran/bloqs/arithmetic/sorting_test.py [hunk 5]
```diff
@@ -27,19 +30,54 @@ def test_bitonic_sort(bloq_autotester):
...
+
+    _ = bloq.t_complexity()
+
+
+def test_bitonic_sort_classical_sim():
+    L = 8
+    xs = np.array([4, 2, 7, 1])
+    bloq = BitonicSort(L, len(xs))
+    sorted_xs, _ = bloq.call_classically(xs=xs)
+    assert np.all(sorted_xs == sorted(xs))
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting_test.py:67]: sweet

File: qualtran/bloqs/arithmetic/sorting.py [hunk 6]
```diff
@@ -11,118 +11,305 @@
...
     return cmp_symb
 
 
 _COMPARATOR_DOC = BloqDocSpec(bloq_cls=Comparator, examples=[_cmp_symb])
 
 
 @frozen
-class BitonicSort(Bloq):
-    r"""Sort k n-bit numbers.
+class ParallelComparators(Bloq):
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting.py:105]: Add unit tests for this and Bitonic merge, we may want to split this file which now has several bloqs, which sort of strays from the ~ 1 bloq / file ideal / WDYT @mpharrigan

File: qualtran/bloqs/arithmetic/sorting.py [hunk 7]
```diff
@@ -11,118 +11,318 @@
...
+
+        junk = np.concatenate(all_junks)
+        assert len(junk) == self.num_comparisons
+
+        return {'result': result, 'junk': np.array(junk)}
+
+
+@frozen
+class BitonicSort(Bloq):
+    r"""Sort k numbers in the range {0, ..., L-1}.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:247]: Use brackets `[0, L)` or `[0, L]` instead of `{0, ..., L - 1}` for consistency. 

```suggestion
    r"""Sort k numbers in the range [0, ..., L-1].
```

File: qualtran/bloqs/arithmetic/sorting.py [hunk 8]
```diff
@@ -11,118 +11,305 @@
...
 
     Args:
-        bitsize: Number of bits used to represent each integer.
+        L: Upper-bound (excluded) on the input integers.
         k: Number of integers to sort.
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting.py:251]: we're sort of using k to mean several different things here (each half, whole, list, ...), maybe a more descriptive variable name is in order.

File: qualtran/bloqs/arithmetic/sorting.py [hunk 9]
```diff
@@ -11,118 +11,305 @@
...
-    bitsize: SymbolicInt
-    k: int
+    L: SymbolicInt
+    k: SymbolicInt
+
+    def __attrs_post_init__(self):
+        k = self.k
+        if not is_symbolic(k):
+            assert k >= 1, "length of input list must be positive"
+            # TODO support non-power-of-two input lengths
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting.py:269]: Open an issue and link it here.

File: qualtran/bloqs/arithmetic/sorting.py [hunk 10]
```diff
@@ -11,118 +11,305 @@
...
+        k: Number of integers in each half
+    """
+    L: SymbolicInt
+    k: SymbolicInt
+
+    def __attrs_post_init__(self):
+        k = self.k
+        if not is_symbolic(k):
+            assert k >= 1, "length of input lists must be positive"
+            # TODO support non-power-of-two input lengths
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting.py:182]: Add link to issue here too.

File: qualtran/bloqs/arithmetic/sorting_test.py [hunk 11]
```diff
@@ -27,19 +30,54 @@ def test_bitonic_sort(bloq_autotester):
...
+
+
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/sorting_test.py:69]: your new bloqs should add autotesters and bloq-doc-specs too.

**Review by fdmalone — CHANGES_REQUESTED** [review]
So nice! I'm glad someone finally did this.

File: qualtran/bloqs/arithmetic/sorting.ipynb (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic/sorting.ipynb:39]: Is there an advantage in terms of complexity of sorting? I would say this notation is definitely more confusing. I had to look at the unit tests to figure out what you mean by `Sort k numbers in the range {0, ..., L-1}.` My first guess was maybe you want to sort the first `k` of the first `L` numbers or something. 

Saying something like `Sort K n-bit numbers in complexity `O()` complexity` is a lot cleaner. If there's no advantage in terms of complexity and everywhere we are using `n=bitsize(L - 1)` to determine the bitsize anyway, then I'd say lets change the bloqs to use `bitsize` of the numbers to be sorted as the attribute by default and we can provide a builder method that expects an `L` and computes the bitsize for you.

File: qualtran/bloqs/arithmetic/sorting.py [hunk 12]
```diff
@@ -11,118 +11,318 @@
...
 _COMPARATOR_DOC = BloqDocSpec(bloq_cls=Comparator, examples=[_cmp_symb])
 
 
 @frozen
-class BitonicSort(Bloq):
-    r"""Sort k n-bit numbers.
+class ParallelComparators(Bloq):
+    """Given k numbers in [0, L), compare every pair that is `offset` apart.
+
+    Args:
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:108]: Please more details to the docstring - do we compare and swap or we just store the result of the comparison somewhere? Are the RIGHT junk registers storing the result of the comparison? Or are they supposed to be completely ignored?

File: qualtran/bloqs/arithmetic/sorting.py [hunk 13]
```diff
@@ -11,118 +11,318 @@
...
+
+
+@frozen
+class BitonicMerge(Bloq):
+    r"""Merge two sorted sequences of integers.
 
     Args:
-        bitsize: Number of bits used to represent each integer.
+        L: Upper-bound (excluded) on the input integers.
+        k: Number of integers in each half
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:172]: Please specify that `k` must be a power of `2`. Also specify the complexity in terms of `T / Toffoli costs`, `ancilla cost` and `circuit depth`. 

Also, would it be a lot more work if we want to support merging two lists of different sizes?

File: qualtran/bloqs/arithmetic/sorting.py [hunk 14]
```diff
@@ -11,118 +11,318 @@
...
+
+        return {'result': result, 'junk': np.array(junk)}
+
+
+@frozen
+class BitonicSort(Bloq):
+    r"""Sort k numbers in the range {0, ..., L-1}.
+
+    Args:
+        L: Upper-bound (excluded) on the input integers.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:250]: Please specify that `k` must be a power of `2`. Also specify the complexity in terms of `T / Toffoli costs`, `ancilla cost` and `circuit depth`.

File: qualtran/bloqs/arithmetic/sorting.py [hunk 15]
```diff
@@ -11,118 +11,318 @@
...
+            # TODO support non-power-of-two input lengths
+            assert (k & (k - 1)) == 0, "length of input list must be a power of 2"
 
     @property
     def signature(self):
-        return Signature([Register("input", QAny(self.bitsize), shape=(self.k,))])
+        return Signature(
+            [
+                Register("xs", BoundedQUInt(self.bitsize, self.L), shape=(self.k,)),
+                Register("junk", QBit(), shape=(self.num_comparisons,), side=Side.RIGHT),
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:277]: Why do we have `junk` RIGHT registers everywhere ? Are we saving on any uncomputation? Can you also expand the docstring to explain this?

File: qualtran/bloqs/arithmetic/sorting_test.py [hunk 16]
```diff
@@ -27,19 +30,54 @@ def test_bitonic_sort(bloq_autotester):
...
-    with pytest.raises(NotImplementedError):
-        bloq.decompose_bloq()
+
+    bloq = BitonicSort(2**bitsize, k)
+    assert bloq.num_comparisons == 24
+
+    _ = bloq.t_complexity()
+
+
+def test_bitonic_sort_classical_sim():
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting_test.py:62]: Would be nice to have a test that tests sorting permutations for all `n!` classical inputs for a list of size `n`. This makes sure we are not missing any corner case that wouldn't get sorted as expected.

File: qualtran/bloqs/arithmetic/sorting_test.py [hunk 17]
```diff
@@ -27,19 +30,54 @@ def test_bitonic_sort(bloq_autotester):
...
+    assert bloq.t_complexity() == TComplexity(t=15 * bitsize + 4, clifford=56 * bitsize + 24)
+
+
+@pytest.mark.parametrize("L", [5, 8, 12])
+def test_comparator_classical_sim(L: int):
+    bloq = Comparator(L)
+    for a in range(L):
+        for b in range(L):
+            res_a, res_b, anc = bloq.call_classically(a=a, b=b)
+            assert res_a <= res_b
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting_test.py:49]: Also assert that `res_a` and `res_b` are equal to `min(a, b)` and `max(a,b)` and `anc` stores the result of the comparison.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
+1, great to have this implemented!

File: qualtran/bloqs/arithmetic/sorting.py (hunk 7 above)
Comment (fdmalone) [reply qualtran/bloqs/arithmetic/sorting.py:247]: Should definitely be [0, L) for exclusive upper limit

File: qualtran/bloqs/arithmetic/sorting.py (hunk 6 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py:105]: So `BitonicSort` is the only end bloq that algorithms use. The rest are just helpers to make the decomposition more modular, I feel they really can't be used in general, other than as sub-steps of `BitonicSort`.

If the intermediate bloqs are more useful, we could split them up into their own files.

Comment (fdmalone) [reply qualtran/bloqs/arithmetic/sorting.py:105]: Ok, I think they should still be unit tested if they have implementations though.

File: qualtran/bloqs/arithmetic/sorting.ipynb (hunk 2 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.ipynb:39]: Actually, as the comparison bloqs only accept bitsize, this doesn't give any improvement. It does make more sense to just keep it as bitsize, (and provide a helper constructor that takes `L` ). I'll revert the changes and update the docs.

File: qualtran/bloqs/arithmetic/sorting.py (hunk 6 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py:105]: Yep will add unit tests for each

File: qualtran/bloqs/arithmetic/sorting.py (hunk 7 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: used bitsizes instead

File: qualtran/bloqs/arithmetic/sorting.ipynb (hunk 2 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.ipynb]: Used bitsizes everywhere.

File: qualtran/bloqs/arithmetic/sorting_test.py (hunk 17 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting_test.py:60]: added the three assertions

File: qualtran/bloqs/arithmetic/sorting_test.py (hunk 16 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting_test.py:76]: added a testing all permutations of `[1,2,3,4]`

File: qualtran/bloqs/arithmetic/sorting.py (hunk 15 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py:269]: This is the comparator circuit from the [reference](https://www.nature.com/articles/s41534-018-0071-5):
image

I wrote the signature similar to the `And` bloq where the result bit is `RIGHT`.

File: qualtran/bloqs/arithmetic/sorting_test.py (hunk 4 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting_test.py:44]: I had some reason to justify this, but can't remember anymore. I _think_ somewhere in the previous t_complexity it assumed CSwap was `t=14`, but the current impl is `t=7`

Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic/sorting_test.py:44]: The paper assumes CSWAP is `t=4` using 1 ancilla, right? We should probably open an issue to track that though. Textbook CSWAP should be `t=7`. 


CSWAP to 1 Toffoli can be reduced as follows
image


and then the Toffoli can be implemented using 4 T either using a Relative phase toffoli or using an ancilla with AND gate.

File: qualtran/bloqs/arithmetic/sorting.py (hunk 15 above)
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic/sorting.py:269]: I think its worth highlighting this in the docstring -- i.e. the junk stores the result of the comparisons. This would be necessary to make the sorting bloq reversible so you can apply the unitary in reverse and get back to the original sequence of unsorted elements if needed.

File: qualtran/bloqs/arithmetic/sorting.py (hunk 9 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: done

File: qualtran/bloqs/arithmetic/sorting.py (hunk 10 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: done

File: qualtran/bloqs/arithmetic/sorting.py (hunk 12 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: updated all docstrings.

File: qualtran/bloqs/arithmetic/sorting.py (hunk 15 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py:324]: updated the docstring explaining this.

File: qualtran/bloqs/arithmetic/sorting.py (hunk 13 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: It isn't too difficult, but the edge cases need to be dealt with care. I wanted to get the power-of-2 case working first, will implement the general case in a follow up.

File: qualtran/bloqs/arithmetic/sorting.py (hunk 8 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py:297]: updated to `half_length`

File: qualtran/bloqs/arithmetic/sorting.py (hunk 6 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py:105]: added simulation tests for each supporting bloq

File: qualtran/bloqs/arithmetic/sorting_test.py (hunk 11 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting_test.py:119]: added bloq_examples with autotesters and updated the bloqdocspecs

File: qualtran/bloqs/arithmetic/sorting.py (hunk 14 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: added to docstring

File: qualtran/bloqs/arithmetic/sorting.py [hunk 18]
```diff
@@ -48,82 +64,331 @@ class Comparator(Bloq):
...
+class BitonicMerge(Bloq):
+    r"""Merge two sorted sequences of n-bit integers.
+
+    Given two sorted lists of length `half_length`, merges them inplace into a single
+    sorted list.
+
+    Currently only supports `half_length` equal to a power of two (#1090).
+
+    If each half has length $k$, then the merge network uses $k * (\log{k} + 1)$ comparisons
+    when $k$ is a power of 2.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:200]: Please also add information / $$..$$ formula to specify what the `junk` stores since it can be potentially useful depending upon the context.

File: qualtran/bloqs/arithmetic/sorting.py [hunk 19]
```diff
@@ -48,82 +64,331 @@ class Comparator(Bloq):
...
+
+_COMPARATOR_DOC = BloqDocSpec(bloq_cls=Comparator, examples=[_comparator, _comparator_symb])
+
+
+@frozen
+class ParallelComparators(Bloq):
+    """Given k n-bit numbers, for each pair that is `offset` apart, compare and swap if needed to make them ordered.
+
+    For each block of `2 * offset` numbers, apply a `Comparator` between each pair that is `offset` apart.
+    Uses up to $k / 2$ Comparators.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:109]: Please add a short note about what the junk stores

File: qualtran/bloqs/arithmetic/sorting.py [hunk 20]
```diff
@@ -48,82 +64,331 @@ class Comparator(Bloq):
...
-        for the moment. Issue #219
+    where $y_1, y_2, \ldots, y_k = \mathrm{sorted}(x_1, x_2, \ldots, x_k$, and the junk register
+    stores the result of comparisons done during the sorting. Note that the junk register will
+    be entangled with the input list register.
+
+    Currently only supports $k$ being a power of two (#1090).
+
+    The bitonic sorting network requires $\frac{k}{2} \frac{\log{k} (\log{k} + 1)}{2}$ total comparisons,
+    and has depth $\frac{\log{k} (\log{k} + 1)}{2}$, when $k$ is a power of 2. Each comparison generates
+    one ancilla qubit, so the total size of junk register equals the number of comparisons.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:309]: This is great!

File: qualtran/serialization/sympy.py [hunk 21]
```diff
@@ -25,6 +25,8 @@ def _get_sympy_function_type(expr: sympy.Basic) -> int:
 
     This method converts a sympy function to its sympy_pb2.Function enum representation.
     """
+    import sympy.codegen.cfunctions
+
```
Comment (tanujkhattar) [inline qualtran/serialization/sympy.py:29]: Why the import? We don't seem to be using it?

File: qualtran/serialization/sympy.py [hunk 22]
```diff
@@ -62,6 +64,8 @@ def _get_sympy_function_from_enum(enum: int) -> Any:
     Sympy functions are represented as a sympy_pb2.Function enum. This method converts
     this int enum.
     """
+    import sympy.codegen.cfunctions
+
```
Comment (tanujkhattar) [inline qualtran/serialization/sympy.py:68]: ditto

**Review by tanujkhattar — APPROVED** [review]
LGTM % some final comments

File: qualtran/serialization/sympy.py (hunk 21 above)
Comment (anurudhp) [reply qualtran/serialization/sympy.py:29]: I added `sympy.codegen.cfunctions.log2` in https://github.com/quantumlib/Qualtran/pull/1088, and the tests passed because `sympy_test.py` imported it. But otherwise `import sympy` itself does not load `sympy.codegen`.

**Review by fdmalone — APPROVED** [review]

File: qualtran/bloqs/arithmetic/sorting.py (hunk 19 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: updated

File: qualtran/bloqs/arithmetic/sorting.py (hunk 18 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: Added a brief description (similar to the one in `BitonicSort`).

Did you also want a description of the order of the ancillas?

File: qualtran/bloqs/arithmetic/sorting.py [hunk 23]
```diff
@@ -48,82 +64,342 @@ class Comparator(Bloq):
...
+    For each block of `2 * offset` numbers, apply a `Comparator` between each pair that is `offset` apart.
+    For an offset of $\delta$, this requires
+
+    $$
+        \delta \left\lfloor \frac{k}{2\delta} \right\rfloor
+        + \max((k\mod 2\delta) - \delta, 0)
+    $$
+
+    The above expression is at most $k / 2$.
+    Each comparison generates one ancilla qubit, which are aggregated into the `junk` register.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:117]:
```suggestion
    Each comparison generates one ancilla qubit which stores the result of the comparison. The ancillas 
    are aggregated into the `junk` register.
```

File: qualtran/bloqs/arithmetic/sorting.py [hunk 24]
```diff
@@ -48,82 +64,342 @@ class Comparator(Bloq):
...
+
+    For each block of `2 * offset` numbers, apply a `Comparator` between each pair that is `offset` apart.
+    For an offset of $\delta$, this requires
+
+    $$
+        \delta \left\lfloor \frac{k}{2\delta} \right\rfloor
+        + \max((k\mod 2\delta) - \delta, 0)
+    $$
+
+    The above expression is at most $k / 2$.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:116]: There seems to be a break after "this requires, <math expression> . The above expression".

We should finish the "this requires," line. Maybe something like
```suggestion
    number of comparisons. The above expression is at most $k / 2$.
```

File: qualtran/bloqs/arithmetic/sorting.py [hunk 25]
```diff
@@ -48,82 +64,342 @@ class Comparator(Bloq):
...
+    r"""Merge two sorted sequences of n-bit integers.
+
+    Given two sorted lists of length `half_length`, merges them inplace into a single
+    sorted list.
+
+    Currently only supports `half_length` equal to a power of two (#1090).
+
+    If each half has length $k$, then the merge network uses $k (1+\log{k})$ comparisons
+    when $k$ is a power of 2.
+    Each comparison generates one ancilla qubit, which are aggregated into the `junk` register.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/sorting.py:209]: ditto: Let's specify that the ancilla qubit stores the result of comparisons

**Review by tanujkhattar — APPROVED** [review]

File: qualtran/bloqs/arithmetic/sorting.py (hunk 18 above)
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic/sorting.py]: The order is not as important, but we should at least specify that the ancilla qubits store the results of intermediate comparisons.

File: qualtran/bloqs/arithmetic/sorting.py (hunk 24 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: oops, fixed

File: qualtran/bloqs/arithmetic/sorting.py (hunk 25 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: updated

File: qualtran/bloqs/arithmetic/sorting.py (hunk 23 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/sorting.py]: updated
